TCAT/Services/Network/NetworkManager.swift

```diff
@@ -0,0 +1,73 @@
+//
+//  NetworkManager.swift
+//  TCAT
+//
+//  Created by Jayson Hahn on 9/15/24.
+//  Copyright © 2024 Cornell AppDev. All rights reserved.
+//
+
+import Foundation
+import Combine
+
+protocol NetworkService {
+    func request<T: Decodable>(_ request: URLRequest, decodingType: T.Type) -> AnyPublisher<T, ApiErrorHandler>
+}
+
+class NetworkManager: NetworkService {
+    
+    private let session: URLSession
+    
+    init(session: URLSession = .shared) {
+        self.session = session
+    }
+    
+    func request<T: Decodable>(_ request: URLRequest, decodingType: T.Type) -> AnyPublisher<T, ApiErrorHandler> {
+        return session.dataTaskPublisher(for: request)
+            .tryMap { result in
+                try self.handleResponse(result)
+            }
+            .decode(type: APIResponse<T>.self, decoder: JSONDecoder())
+            .tryMap { response in
+                try self.validateAPIResponse(response)
+            }
+            .mapError { error in
+                self.mapToAPIError(error)
+            }
+            .eraseToAnyPublisher()
+    }
+    
+    // Handles HTTP response and decodes or throws an appropriate error
+    private func handleResponse(_ result: URLSession.DataTaskPublisher.Output) throws -> Data {
+        guard let httpResponse = result.response as? HTTPURLResponse else {
+            throw ApiErrorHandler.requestFailed
+        }
+        if (200..<300).contains(httpResponse.statusCode) {
+            return result.data
+        } else {
+            // Attempt to decode error message from server
+            if let apiError = try? JSONDecoder().decode(ApiError.self, from: result.data) {
+                throw ApiErrorHandler.customApiError(apiError)
+            } else {
+                throw ApiErrorHandler.emptyErrorWithStatusCode(httpResponse.statusCode.description)
+            }
+        }
+    }
+    
+    // Validate API response and handle future error cases
+    private func validateAPIResponse<T>(_ response: APIResponse<T>) throws -> T {
+        guard response.success else {
+            // This is a placeholder error handler. Update as needed. When backend sends more error codes.
```

Comment:
Add a TODO comment in here so it doesn't get lost in the sauce when you need to update it? Doesn't impact any code though soo